# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: mfbt/tests/TestBufferList.cpp
# Commit: 451508713a62
# Full Hash: 451508713a625beaaa9fcbbf06c65ca312dc34fb
# Author: Nika Layzell <nika@thelayzells.com>
# Date: 2021-06-23 09:53:24
# Regressor Bug: 1706374
# File Overlap Count: 2
# Description:
#   Bug 1706374 - Part 13: Remove the event footer from an IPC::Message when deserializing, r=handyman,glandium
#   
#   This unfortunately requires a new method to be added to BufferList to
#   support truncating the buffer to a particular iterator.
#   
# ==============================================================================

diff -r 878722e87e62 -r 451508713a62 mfbt/tests/TestBufferList.cpp
--- a/mfbt/tests/TestBufferList.cpp	Tue Jun 22 18:17:23 2021 +0000
+++ b/mfbt/tests/TestBufferList.cpp	Tue Jun 22 18:17:23 2021 +0000
@@ -367,5 +367,67 @@
   MOZ_RELEASE_ASSERT(iter1.AdvanceAcrossSegments(bl12, 1));
   MOZ_RELEASE_ASSERT(iter1.Done());
 
+  BufferList bl13(0, 0, 8);
+  MOZ_ALWAYS_TRUE(bl13.WriteBytes("abcdefgh", 8));
+  MOZ_ALWAYS_TRUE(bl13.WriteBytes("12345678", 8));
+  MOZ_ALWAYS_TRUE(bl13.WriteBytes("ABCDEFGH", 8));
+  MOZ_RELEASE_ASSERT(bl13.Size() == 24);
+
+  // At segment border
+  iter = bl13.Iter();
+  MOZ_RELEASE_ASSERT(iter.AdvanceAcrossSegments(bl13, 8));
+  MOZ_RELEASE_ASSERT(bl13.Truncate(iter) == 16);
+  MOZ_RELEASE_ASSERT(iter.Done());
+  MOZ_RELEASE_ASSERT(bl13.Size() == 8);
+
+  // Restore state
+  MOZ_ALWAYS_TRUE(bl13.WriteBytes("12345678", 8));
+  MOZ_ALWAYS_TRUE(bl13.WriteBytes("ABCDEFGH", 8));
+  MOZ_RELEASE_ASSERT(bl13.Size() == 24);
+
+  // Before segment border
+  iter = bl13.Iter();
+  MOZ_RELEASE_ASSERT(iter.AdvanceAcrossSegments(bl13, 7));
+  MOZ_RELEASE_ASSERT(bl13.Truncate(iter) == 17);
+  MOZ_RELEASE_ASSERT(iter.Done());
+  MOZ_RELEASE_ASSERT(bl13.Size() == 7);
+
+  // Restore state
+  MOZ_ALWAYS_TRUE(bl13.WriteBytes("h", 1));
+  MOZ_ALWAYS_TRUE(bl13.WriteBytes("12345678", 8));
+  MOZ_ALWAYS_TRUE(bl13.WriteBytes("ABCDEFGH", 8));
+  MOZ_RELEASE_ASSERT(bl13.Size() == 24);
+
+  // In last segment
+  iter = bl13.Iter();
+  MOZ_RELEASE_ASSERT(iter.AdvanceAcrossSegments(bl13, 20));
+  MOZ_RELEASE_ASSERT(bl13.Truncate(iter) == 4);
+  MOZ_RELEASE_ASSERT(iter.Done());
+  MOZ_RELEASE_ASSERT(bl13.Size() == 20);
+
+  // No-op truncate
+  MOZ_RELEASE_ASSERT(bl13.Truncate(iter) == 0);
+  MOZ_RELEASE_ASSERT(iter.Done());
+  MOZ_RELEASE_ASSERT(bl13.Size() == 20);
+
+  // No-op truncate with fresh iterator
+  iter = bl13.Iter();
+  MOZ_RELEASE_ASSERT(iter.AdvanceAcrossSegments(bl13, 20));
+  MOZ_RELEASE_ASSERT(bl13.Truncate(iter) == 0);
+  MOZ_RELEASE_ASSERT(iter.Done());
+  MOZ_RELEASE_ASSERT(bl13.Size() == 20);
+
+  // Truncate at start of buffer
+  iter = bl13.Iter();
+  MOZ_RELEASE_ASSERT(bl13.Truncate(iter) == 20);
+  MOZ_RELEASE_ASSERT(iter.Done());
+  MOZ_RELEASE_ASSERT(bl13.Size() == 0);
+
+  // No-op truncate at start of buffer
+  iter = bl13.Iter();
+  MOZ_RELEASE_ASSERT(bl13.Truncate(iter) == 0);
+  MOZ_RELEASE_ASSERT(iter.Done());
+  MOZ_RELEASE_ASSERT(bl13.Size() == 0);
+
   return 0;
 }
